### PR TITLE
Allow calling 'magic' methods inside 'magic' methods.

### LINKED ISSUE #2381: 609: Allow __enter__() inside an __enter__()
### What's wrong

One design pattern is to wrap a context manager. It would be nice to avoid WPS609 errors with this code, which seems to require accessing the direct magic methods.

### How it should be

Allow code like:
```
class Foo:
    ...

    def __enter__(self):
        self._conn.__enter__()
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self._conn.__exit__(exc_type, exc_value, traceback)
```

I guess the same for aenter/aexit as well.

### PR DESCRIPTION
**Pull request by lensvol**
Resolves #2381.

# I have made things!

## Checklist

- [X] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [X] I have created at least one test case for the changes I have made
- [X] I have updated the documentation for the changes I have made
- [X] I have added my changes to the `CHANGELOG.md`

## Related issues

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py [hunk 1]
```diff
@@ -107,6 +108,35 @@ def method(cls):
...
+
+    def something(cls):
+        class Test(object):
+            def {0}(cls):
+                self.conn.{0}()
+
+        anything()
+"""
+
+magic_wrapper_method_with_stacked_wrapper_cls = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py:129]: Is this used?

File: wemake_python_styleguide/visitors/ast/attributes.py [hunk 2]
```diff
@@ -24,6 +30,19 @@ class WrongAttributeVisitor(BaseNodeVisitor):
         'mcs',
     ))
 
+    _func_stack: List[str] = []
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/attributes.py:33]: Any reason to use a class var?

Comment (sobolevn) [reply wemake_python_styleguide/visitors/ast/attributes.py:33]: I think we can get context nodes and see if we are in a magic method that is inside a class.

File: tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py (hunk 1 above)
Comment (lensvol) [reply tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py]: No, removed it in a separate commit.

File: wemake_python_styleguide/visitors/ast/attributes.py (hunk 2 above)
Comment (lensvol) [reply wemake_python_styleguide/visitors/ast/attributes.py:33]: Interesting, I will try reworking using that.

File: tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py [hunk 3]
```diff
@@ -107,6 +108,24 @@ def method(cls):
         super().{0}()
 """
 
+magic_wrapper_method = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py:111]: Should we allow cases like:

```python
def __init__(...):
    other.__init__(...)
```

?

I think it makes sense for `__getattr__` and `__dir__` which can be used on a module level. But, not for anything else. Do you agree?

**Comment by lensvol** [comment]
@sobolevn I've rewritten it using context, thanks for the suggestion!

File: tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py (hunk 3 above)
Comment (lensvol) [reply tests/test_visitors/test_ast/test_attributes/test_magic_attributes.py:111]: Do you want to allow only a limited subset of _magic_ methods here? With an explicitly provided list?

I do not see how that can significantly improve the utility of this feature.

File: wemake_python_styleguide/visitors/ast/attributes.py [hunk 4]
```diff
@@ -50,6 +51,15 @@ def _check_protected_attribute(self, node: ast.Attribute) -> None:
 
     def _check_magic_attribute(self, node: ast.Attribute) -> None:
         if access.is_magic(node.attr):
+            # If "magic" method being called has the same name as
+            # the enclosing function, then it is a "wrapper" and thus
+            # a "false positive".
+
+            ctx = nodes.get_context(node)
+            if isinstance(ctx, (ast.FunctionDef, ast.AsyncFunctionDef)):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/attributes.py:59]: Use `from wemake_python_styleguide.compat.aliases import FunctionNodes`

**Review by sobolevn — COMMENTED** [review]
Ok, fair enough! Only one comment left :)

File: wemake_python_styleguide/visitors/ast/attributes.py (hunk 4 above)
Comment (lensvol) [reply wemake_python_styleguide/visitors/ast/attributes.py]: See new commit.

**Review by sobolevn — APPROVED** [review]
Thanks!

**Pull request merged**